:memo: Updated Getting started / Configuration (#5)
writechoiceorg/terrateam: docs/src/content/docs/getting-started/configuration.mdx

---

> title: Configuration 

Reviewer: A "Basic Example" of a complete config.yml file at the end would help users see how the sections work together.

Author: done

description: Understanding the Terrateam configuration file and how to customize your workflow
---

> import { Steps } from '@astrojs/starlight/components';

Reviewer 2: I can't add comments to the following paragraph. For this reason I adding it here.

Please, explain how the file is generated. Is it generated when you install the GitHub app? Or should the user add it by themselves?

Reviewer 2: In addition, add the following note here as well:

:::note
Default settings work for most teams. Customize only if you need specific lock timing or policies.
:::

```yaml
access_control:
  # Specifies user permissions for operations
apply_requirements:
  # Defines preconditions for applying changes
dirs:
  # Maps directories to tags, workspaces, and behaviors
hooks:
  # Executes custom commands before or after operations
workflows:
  # Configures specialized plan and apply processes
```

Each section serves a specific purpose in customizing Terrateam's behavior:

- **access_control**: Controls who can run plan and apply operations
- **apply_requirements**: Defines conditions that must be met before changes can be applied
- **dirs**: Maps your repository's directory structure to Terrateam's concepts
- **hooks**: Allows custom actions before or after operations
- **workflows**: Creates specialized processing for different directories or environments

> ## Understanding Tag Queries

Reviewer 2: Why do we have this content about Tag here @lucaspr2 .
It looks like it is from other page.

Author: @Arcapas suggested a introduction to this concept, i think its relevant to understanding the page, but i can remove it if you think its not needed. @htessaro

:::note
The `tag_query` field is used to match specific directories or environments. An empty string (`''`) matches all directories. You can use it to create different access policies for different parts of your infrastructure, like `tag_query: 'production'` to set stricter permissions for production environments.
:::

For more information, see the [Access Control](/configuration-reference/access-control) section of the Configuration Reference.

## Apply Requirements

Apply Requirements allows you to specify conditions that must be met before an Apply operation can be performed on an unmerged pull request. This helps ensure that changes are properly reviewed and validated before being applied.

Here's an example `apply_requirements` configuration with explanations:

```yaml
apply_requirements:
  # Creates a status check that prevents merging until apply is complete
  create_pending_apply_check: true
  
  checks:
    - tag_query: ""  # Apply to all directories
      
      # Requires PR approvals
      approved:
        enabled: true        # Enable approval checking
        any_of: []           # No specific approvers required

>         any_of_count: 2      # Requires at least 2 approvals

Reviewer: ```suggestion
        any_of_count: 2      # Requires at least two approvals
```

        all_of: []           # No mandatory approvers
      
      # Prevents apply if merge conflicts exist
      merge_conflicts:
        enabled: true
      
      # Requires all status checks to pass
      status_checks:
        enabled: true              # Enable status check verification
        ignore_matching:           # Regex patterns for checks to ignore
          - "ci/.*"                # Ignores all CI status checks
```

> This configuration requires that the pull request has at least 2 approvals, no merge conflicts, and all status checks (except those matching `ci/.*`) have passed before an apply can be performed.

Reviewer: ```suggestion
This configuration requires that the pull request has at least two approvals, no merge conflicts, and all status checks (except those matching `ci/.*`) have passed before an apply can be performed.
```

```yaml
access_control:
  policies:
    - tag_query: ''  # Empty string matches all directories
      plan: ['*']    # Anyone can plan
      apply: ['team:sre']  # Only SRE team members can apply
```

This configuration allows anyone to trigger a Plan operation but restricts Apply operations to members of the `sre` team.

> See [Access Control](/configuration-reference/access-control) for details.

Reviewer 2: ```suggestion
Access the [Access Control](/configuration-reference/access-control) page to see the complete list of configurations available. 
```

Here's an example `dirs` configuration:

```yaml
dirs:
  ec2:  # Directory name (relative to repository root)
    tags: [aws, ec2]  # Directory-level tags
    workspaces:
      production:  # Workspace name
        tags: [production]  # Workspace-specific tags
    when_modified:
      file_patterns: ["ec2/*.tf", "ec2/*.tfvars", "iam/*.tf", "iam/*.tfvars"]
  iam:
    tags: [aws, iam]
```

This configuration:

1. Assigns directory-level tags `aws` and `ec2` to the `ec2` directory
2. Assigns directory-level tags `aws` and `iam` to the `iam` directory
3. Creates a `production` workspace for the `ec2` directory with the tag `production`
4. Specifies custom file patterns that, when modified, will trigger Terrateam operations for the `ec2` directory

:::note[Directory vs. Workspace Tags]
- **Directory tags** (`tags` at the directory level) apply to all operations in that directory across all workspaces
- **Workspace tags** (`tags` under a specific workspace) only apply when operating on that specific workspace

> - Tags are used in tag queries (like `tag_query: "aws"`) to target specific operations

Reviewer: Tag queries might benefit from a clearer explanation. If feasible, consider placing an introductory overview earlier.

> 

Reviewer: The "Workflows" section features an advanced example. It would benefit from a simpler introductory example before exploring customization.

```yaml
workflows:
  # This workflow applies to any directory tagged with "production"
  # (This could match directories set with tags: [production] in the dirs section)
  - tag_query: "production"
    plan:
      - type: init  # Run terraform init
      - type: plan  # Run terraform plan
        extra_args: ["-var-file=production.tfvars"]  # With these extra arguments
    apply:
      - type: init  # Run terraform init for apply
      - type: apply  # Run terraform apply
      - type: run   # Run a custom command
        cmd: ['echo', 'Error running apply']
        run_on: failure  # Only if apply fails
```

This workflow runs a custom plan and apply process for any directories tagged with `production`. The tag queries like `tag_query: "production"` match the tags defined in the `dirs` section, creating a connection between your directory structure and custom workflows.

:::note
All the examples in this page are complementary parts of a single configuration. The tag query in the workflow (`tag_query: "production"`) would match directories or workspaces tagged with `production` in the `dirs` section.
:::

> For more information, see the [Workflows](/configuration-reference/workflows) section of the Configuration Reference.

Reviewer: "For more information,..." – Repeats in multiple sections; use "See [section] for details" instead.

```yaml
hooks:
  all:  # Apply to all operations (plan & apply)
    pre:  # Run before the operation
      - type: run
        cmd: ['echo', 'Running pre-hook for all operations']
  plan:  # Apply only to plan operations
    post:  # Run after the operation
      - type: run
        cmd: ['echo', 'Running post-hook for plan operations']
```

See [Hooks](/configuration-reference/hooks) for details.

## Workflows

Workflows allow you to define custom steps for Terrateam's Plan and Apply operations. You can use Workflows to replace or augment the default behavior.

> ### Simple Workflow Example

Reviewer 2: ```suggestion
```

Here's a basic workflow that slightly modifies the default behavior:

```yaml
workflows:
  # Apply to all directories
  - tag_query: ""
    plan:
      - type: init  # Run terraform init
      - type: plan  # Run terraform plan
    apply:
      - type: init  # Run terraform init
      - type: apply # Run terraform apply
```

> This simple workflow explicitly defines the standard Terraform workflow steps.

Reviewer 2: Please, provide a explanation about what happens when this workflows is executed.

> ### Advanced Workflow Example

Reviewer 2: ```suggestion
```

> For more complex scenarios, you can create specialized workflows:

Reviewer 2: ```suggestion
For more complex scenarios, you can create specialized workflows, such as the next one:
```

- `cost_estimation`: Enable and configure cost estimation for Terraform plans
- `automerge`: Automatically merge pull requests after successful applies

See the [Configuration Reference](/configuration-reference) for a complete list of available options.

## Basic Example

> Here's a complete basic example of a `config.yml` file that combines the various sections:

Reviewer 2: ```suggestion
Here's a complete example of a `config.yml` file containing a basic workflow that combines the configurations from the previous sections:
```

```yaml
# Basic Terrateam configuration example
access_control:
  policies:
    - tag_query: ''  # All directories
      plan: ['*']    # Anyone can plan
      apply: ['team:infra', 'team:platform']  # Only specific teams can apply

apply_requirements:
  create_pending_apply_check: true
  checks:
    - tag_query: ""  # All directories
      approved:
        enabled: true
        any_of_count: 1  # Requires at least 1 approval
      
dirs:
  staging:
    tags: [aws, staging]
  production:
    tags: [aws, production, critical]
    workspaces:
      default:
        tags: [default]

hooks:
  all:
    pre:
      - type: run
        cmd: ['echo', 'Starting Terraform operation']

workflows:
  - tag_query: "staging"
    plan:
      - type: init
      - type: plan
        extra_args: ["-var-file=staging.tfvars"]
  
  - tag_query: "production"
    plan:
      - type: init
      - type: plan
        extra_args: ["-var-file=production.tfvars"]
```

> This configuration:

Reviewer 2: Add endpoint to all sentences in the list.

Author: done
